pkg/controller/tas/node_failure_controller.go

```diff
@@ -0,0 +1,224 @@
+	var allWorkloads kueue.WorkloadList
+	if err := r.client.List(ctx, &allWorkloads); err != nil {
+		return nil, fmt.Errorf("failed to list workloads: %w", err)
+	}
+	workloadsToProcess := sets.New[types.NamespacedName]()
+	for _, wl := range allWorkloads.Items {
+		if !isAdmittedByTAS(&wl) {
+			continue
+		}
+		for _, podSetAssignment := range wl.Status.Admission.PodSetAssignments {
+			topologyAssignment := podSetAssignment.TopologyAssignment
+			if topologyAssignment == nil {
+				continue
+			}
+			if !utiltas.IsLowestLevelHostname(topologyAssignment.Levels) {
+				continue
+			}
+			for _, domain := range topologyAssignment.Domains {
+				if nodeName == domain.Values[len(domain.Values)-1] {
+					workloadsToProcess.Insert(types.NamespacedName{Name: wl.Name, Namespace: wl.Namespace})
+				}
+			}
+		}
+	}
+	return workloadsToProcess, nil
+}
+
+// patchWorkloadsForUnavailableNode finds workloads with pods on the specified node
+// and patches their status to indicate the node is unavailable.
+func (r *nodeFailureReconciler) patchWorkloadsForUnavailableNode(ctx context.Context, nodeName string) error {
+	workloadsToProcess, err := r.getWorkloadsOnNode(ctx, nodeName)
+	if err != nil {
+		return err
+	}
+	var workloadProcessingErrors []error
+	for _, wlKey := range workloadsToProcess.UnsortedList() {
+		var wl kueue.Workload
+		if err := r.client.Get(ctx, wlKey, &wl); err != nil {
+			if apierrors.IsNotFound(err) {
+				r.log.V(4).Info("Workload not found, skipping", "workload", wlKey)
+			} else {
+				r.log.V(2).Error(err, "Failed to get workload", "workload", wlKey)
+				workloadProcessingErrors = append(workloadProcessingErrors, err)
+			}
+			continue
+		}
+
+		err := clientutil.Patch(ctx, r.client, &wl, true, func() (bool, error) {
+			currentAnnotations := wl.GetAnnotations()
+			if currentAnnotations == nil {
+				currentAnnotations = make(map[string]string)
+			}
+			existingFailedNode, ok := currentAnnotations[kueuealpha.NodeToReplaceAnnotation]
+			if ok && existingFailedNode == nodeName {
+				return false, nil
+			}
+			currentAnnotations[kueuealpha.NodeToReplaceAnnotation] = nodeName
+			wl.SetAnnotations(currentAnnotations)
+			r.log.V(4).Info("Adding unavailable node to workload annotation", "workload", wlKey, "nodeName", nodeName)
+			return true, nil
+		})
+		if err != nil {
+			r.log.V(2).Error(err, "Failed to patch workload with annotation", "workload", wlKey)
+			workloadProcessingErrors = append(workloadProcessingErrors, err)
+		} else {
+			r.log.V(3).Info("Successfully patched workload with annotation", "workload", wlKey, "unavailableNodesAnnotation", wl.GetAnnotations()[kueuealpha.NodeToReplaceAnnotation])
+		}
+	}
+	if len(workloadProcessingErrors) > 0 {
+		return errors.Join(workloadProcessingErrors...)
+	}
+	return nil
+}
+
+func (r *nodeFailureReconciler) getNodeCondition(node *corev1.Node, conditionType corev1.NodeConditionType) *corev1.NodeCondition {
```

Comment:
nit: can you move it to `pkg/util/node/node.go` (create new file). I know this is the only use so far, but I think it would have been also useful for my merged PR, so I could follow up to use it.

Reply:
Actually, a helper like this could be helpeful already within this PR for integration tests, to simplify or drop setNodeConditionInObject.